src/cacheExchange.ts

```diff
@@ -233,14 +234,15 @@ export const cacheExchange = (opts?: CacheExchangeOpts): Exchange => ({
 
     // Clear old optimistic values from the store
     const { key } = operation;
-    if (optimisticKeys.has(key)) {
+    const hasOptimistic = optimisticKeys.has(key);
+    if (hasOptimistic) {
       optimisticKeys.delete(key);
       clearOptimistic(store.data, key);
     }
 
     let writeDependencies: Set<string> | void;
     let queryDependencies: Set<string> | void;
-    if (data !== null && data !== undefined) {
+    if (data !== null && data !== undefined && !error) {
```

Comment:
This also disables moving into resolvers/updates on erroneous results which I think would allow us to give better typings

Reply:
This is a bug. We shouldn't care _why_ we have no data here.